core/controllers/ingest.py

```diff
@@ -162,72 +146,14 @@ def populate_db(workbook: dict[str, pd.DataFrame], mappings: tuple[DataMapping])
-    programme_exists_same_round = (
-        Programme.query.join(Project)
-        .join(Submission)
-        .filter(Programme.programme_id == programme_id)
-        .filter(Submission.reporting_round == reporting_round)
-        .first()
-    )
-    # if added before or during this round, get the programme entity to merge (update)
-    programme_exists_previous_round = (
-        Programme.query.join(Project)
-        .join(Submission)
-        .filter(Programme.programme_id == programme_id)
-        .filter(Submission.reporting_round <= reporting_round)
-        .first()
-    )
-
-    if programme_exists_same_round:
-        matching_project = None
-        for project in programme_exists_same_round.projects:
-            if project.submission.reporting_round == reporting_round:
-                matching_project = project
-                break
-        if matching_project:
-            # get id of submission to replace and re-use it
-            submission_id = matching_project.submission.submission_id
-            # submission instance to remove
-            submission_to_del = matching_project.submission
-            # drop submission, all children should be dropped via cascade
-            Submission.query.filter_by(id=submission_to_del.id).delete()
-            db.session.flush()
-
-    else:
-        submission_id = next_submission_id(reporting_round)
+    submission_id = get_submission_id(workbook)
 
     for mapping in mappings:
-        worksheet = workbook[mapping.table]
-        if "Submission ID" in mapping.column_mapping:
-            worksheet["Submission ID"] = submission_id
-        models = mapping.map_data_to_models(worksheet)
-
-        if mapping.table == "Programme_Ref" and programme_exists_previous_round:
-            # Set incoming model pk to match existing DB row pk (this record will then be updated).
-            programme_to_merge = models[0]
-            programme_to_merge.id = programme_exists_previous_round.id
-            db.session.merge(programme_to_merge)  # There can only be 1 programme per ingest.
-            continue
-
-        if mapping.table == "Organisation_Ref":
-            organisation_exists = Organisation.query.filter(
-                Organisation.organisation_name == models[0].organisation_name
-            ).first()
-            # If this org already in DB, merge to re-use pk, otherwise use add_all as per loop continuation
-            if organisation_exists:
-                org_to_merge = models[0]
-                org_to_merge.id = organisation_exists.id
-                db.session.merge(org_to_merge)  # There can only be 1 org per ingest.
-                continue
-
-        # for outcome and output dim (ref) data, if record already exists, do nothing.
-        if mapping.table in ["Outputs_Ref", "Outcome_Ref"]:
-            models = get_outcomes_outputs_to_insert(mapping, models)
-
-        db.session.add_all(models)
+        load_function = TABLE_TO_LOAD_FUNCTION[mapping.table]
```

Comment:
how would you envisage this being used for, say, a different fund with a different set of tables to load? Would you have to build a new mapping for each and a new TABLE_TO_LOAD_FUNCTION? Or just one of them?


For instance, a submission with no projects, and only half of the "event" data tables?

Ideally it'd be whatever option makes it the most easily configurable without the need for loads of new code definition, so like a subset of TABLE_TO_LOAD_FUNCTION for each fund type. But then does this seem backwards, as it wwill stil literate through mappings, many of which won't be applicable?

Reply:
Well, I imagine the difference between funds will be that some will not load any data into certain tables. We have the same situation with historical data. Our solution to that has been to iterate through the mapping but to 'jump over' explicitly excluded tables. If we move away from the mapping we may be able to configure this elsewhere.

Reply:
Otherwise something like, ``` if load_function := TABLE_TO_LOAD_FUNCTION.get(mapping.table)```. That way if there is no such table for a specific fund it will be None and the line can be skipped.